core/src/main/java/io/parsingdata/metal/data/Slice.java

```diff
@@ -26,37 +29,44 @@
 
     public final Source source;
     public final long offset;
-    private final byte[] data; // Private because array content is mutable.
-    public final int size;
+    public final BigInteger length;
 
-    public Slice(final Source source, final long offset, final byte[] data) {
+    public Slice(final Source source, final long offset, final BigInteger length) {
         this.source = checkNotNull(source, "source");
         this.offset = offset;
-        this.data = checkNotNull(data, "data");
-        this.size = data.length;
+        this.length = checkNotNull(length, "length");
     }
 
     public byte[] getData() {
-        return data.clone();
+        return getData(length);
+    }
+
+    public byte[] getData(final BigInteger limit) {
+        if (limit.compareTo(BigInteger.ZERO) < 0) { return new byte[] {}; }
+        final BigInteger calculatedLength = limit.compareTo(length) > 0 ? length : limit;
+        try {
+            return source.getData(offset, calculatedLength);
+        } catch(IOException e) {
+            throw new UncheckedIOException(e);
```

Comment:
`Slice` is the lowest common implemented abstraction (there are different sources and `ByteStream` is an interface, so that's why we make the switch from checked to unchecked there.

The reason we switch it is to prevent it from completely spreading across Metal: due to lazyness, it would end up on things like the `eval()` methods of `Expression` and `ValueExpression`. So here we decided to convert it everywhere.

Reply:
It is quite interesting I think. I get your point, though it seems strange that the lowest abstraction throws an unchecked, when higher up is chosen for a checked version and it should be explicitely handled there.

Reply:
First, to prevent me from saying something confusing: the Java API is what I consider the _lowest_ level and the final application using some parser built with Metal is the _highest_ level.

Now, at the lowest level, a checked exception is thrown, which makes sense because problems around IO need to be dealt with because they tend to be fatal (but quite often, fixable). Now, it doesn't really make sense for us to have lots of activity that depends on IO not honor this contract around these `IOException`s.

At the same time, they have no purpose _inside_ Metal: we never fix or even log anything around `IOException`s, because these are always outside scope. They are caused by something outside the control of Metal (physical storage, `ByteStream` implementation, etc.) and they can only be fixed by something outside the control of Metal (hardware restart, user changes target device, etc.)

So inside Metal, all these checked `IOException`s do is just be everywhere and not do _anything_ useful. So that's why I think it makes sense to just take the lowest common implemented IO abstraction (`Slice`) and turn it into an `UncheckedIOException` there, which contains the original `IOException`.

My advice would be for all clients of Metal to catch this explicitly and handle it appropriately.